web3pi_proxy/core/rpc/node/endpoint_pool/pool_manager.py

```diff
@@ -92,14 +90,114 @@ def check_connections(self, pools: List[EndpointConnectionPool]):
-                self.__logger.warning(f"Endpoint `{pool.endpoint.name}`: falling behind")
-
             self.__logger.debug(f"Endpoint `{pool.endpoint.name}`: okay.")
 
 
+class HttpResponseParserListenerSyncController:
+    body: bytes = b''
+    block_number: int = 0
+    block_timestamp: int = 0
+    __logger = get_logger("SyncController")
+
+    def __init__(self, __pool_name: str):
+        super().__init__()
+        self.__pool_name = __pool_name
+
+    def on_body(self, body: bytes):
+        self.body = self.body + body
+
+    def on_message_complete(self):
+        try:
+            block_data = json.loads(self.body)
+            result = block_data.get('result')
+            if not result:
+                error = block_data.get('error')
+                if error:
+                    self.__logger.error(f"{self.__pool_name}: Sync test failed: {error}")
+                else:
+                    self.__logger.error(f"{self.__pool_name}: Sync test failed: reason unknown")
+                return
+            self.block_number = int(result['number'], 16)
+            self.block_timestamp = int(result['timestamp'], 16)
+        except Exception as error:
+            self.__logger.error("%s: %s", error.__class__, error)
+            self.__logger.error(f"{self.__pool_name}: Failed to failed to parse sync test response")
+
+
+# TODO: Tune parameters
+class SyncController:
+    __SUSPENSION_TIMEOUT_SECONDS = 60
+    __MAX_DELAY_SECONDS = 60
+
+    __logger = get_logger("SyncController")
+
+    def __test_pool(self, pool: EndpointConnectionPool) -> bool | None:
+        endpoint_connection_handler: EndpointConnectionHandler | None = None
+        try:
+            endpoint_connection_handler = pool.get(out_of_sync=True)
+            req = RPCRequest()
+            req.headers = b'User-Agent: curl/8.0.1\r\nAccept: */*\r\nContent-Type: application/json\r\nContent-Length: 82\r\n'
+            req.content = b'{"method":"eth_getBlockByNumber","params":["latest",false],"id":1,"jsonrpc":"2.0"}'
+            endpoint_connection_handler.send(req)
+
+            response_listener = HttpResponseParserListenerSyncController(pool.endpoint.name)
+            response_parser = HttpResponseParser(response_listener)
+
+            def response_handler(res: bytes):
+                nonlocal response_parser
+                response_parser.feed_data(res)
+
+            endpoint_connection_handler.receive(response_handler)
+
+            current_timestamp = int(time.time())
+            block_timestamp = response_listener.block_timestamp
+            if block_timestamp == 0:
+                return None  # no logs, failure already logged by the listener
+            if block_timestamp + self.__MAX_DELAY_SECONDS > current_timestamp:
+                return True
+            else:
+                self.__logger.warning(f"{pool.endpoint.name}: The node out of sync")
+                return False
+        except Exception as error:
```

Comment:
especially, if we're catching all errors here anyway, we would be better with just throwing regular exceptions at the listener level...